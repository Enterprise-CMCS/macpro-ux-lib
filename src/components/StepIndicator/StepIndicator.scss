@use "../../assets/theme/macbis-colors" as c;

.usa-step-indicator {
  .usa-step-indicator__segment--current::before,
  .usa-step-indicator__segment--current::after {
    background-color: c.$primary;
  }

  .usa-step-indicator__segment--complete::before,
  .usa-step-indicator__segment--complete::after {
    background-color: c.$primary-darkest;
  }

  .usa-step-indicator__current-step {
    background-color: c.$primary;
  }

  .usa-step-indicator__segment:not(.usa-step-indicator__segment--current, .usa-step-indicator__segment--complete)
    .usa-step-indicator__segment-label {
    color: c.$gray-medium;
  }

  .usa-step-indicator__segment--current .usa-step-indicator__segment-label {
    color: c.$primary;
  }
}
