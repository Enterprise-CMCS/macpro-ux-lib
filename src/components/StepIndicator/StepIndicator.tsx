import React from "react";

type IntrinsicElements = JSX.IntrinsicElements["p"];

type Step = { label: string; order: number };

interface Props extends IntrinsicElements {
  headingText: string;
  currentProgress: number;
  steps: Step[];
  counters?: boolean;
  smallCounters?: boolean;
  showLabels?: boolean;
  isCentered?: boolean;
}

/**
 * **StepIndicator Component**
 *
 * @param {string}     headingText        The heading text of the step indicator.
 * @param {number}     currentProgress    Number tracker of which step the user is currently on.
 * @param {Step[]}     steps              An array of step object data which consist of a label and the order number. The order number is 1 based.
 * @param {boolean}   [counters]          âˆ‚Boolean which shows the counters alternative to the step indidcator.
 * @param {boolean}   [smallCounters]     Boolean which shows the smaller counters alternative to the step indicator. Only shows if counters is true.
 * @param {boolean}   [showLabels]        Dictates whether or not labels are show under each step.
 * @param {boolean}   [isCentered]        Dictates whether the labels are centered under the steps. Labels must be true.
 */
export const StepIndicator: React.FC<Props> = ({
  counters = false,
  smallCounters = false,
  showLabels = true,
  currentProgress,
  steps,
  isCentered = false,
  headingText,
  ...rest
}) => {
  return (
    <div
      className={`usa-step-indicator${
        isCentered ? " usa-step-indicator--center" : ""
      }${
        counters
          ? ` usa-step-indicator--counters${smallCounters ? "-sm" : ""}`
          : ""
      }`}
      aria-label="progress"
      role="progressbar"
      {...rest}
    >
      <ol className="usa-step-indicator__segments">
        {steps.map((step) => {
          return (
            <li
              key={step.label}
              className={`usa-step-indicator__segment ${
                step.order === currentProgress
                  ? "usa-step-indicator__segment--current"
                  : step.order < currentProgress
                  ? "usa-step-indicator__segment--complete"
                  : ""
              }`}
            >
              <span className="usa-step-indicator__segment-label">
                {showLabels && step.label}
                {step.order !== currentProgress && (
                  <span className="usa-sr-only">
                    {step.order > currentProgress
                      ? "completed"
                      : "not completed"}
                  </span>
                )}
              </span>
            </li>
          );
        })}
      </ol>
      <div className="usa-step-indicator__header">
        <h4 className="usa-step-indicator__heading">
          <span className="usa-step-indicator__heading-counter">
            <span className="usa-sr-only">Step</span>
            <span className="usa-step-indicator__current-step margin-right-05">
              {currentProgress}
            </span>
            <span className="usa-step-indicator__total-steps">
              of {steps.length}
            </span>{" "}
          </span>
          <span className="usa-step-indicator__heading-text">
            {headingText}
          </span>
        </h4>
      </div>
    </div>
  );
};
