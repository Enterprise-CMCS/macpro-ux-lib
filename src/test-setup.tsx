import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import "@testing-library/jest-dom";

// Here we can set all providers for themes, stores and other stuff
const AllTheProviders = ({ children }: { children: JSX.Element }) => {
  return <>{children}</>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
