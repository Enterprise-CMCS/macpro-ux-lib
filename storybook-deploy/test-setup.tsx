import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import "@testing-library/jest-dom";

// Here we can set all providers for themes, stores and other stuff
const AllTheProviders = ({ children }: { children: JSX.Element }) => {
  return <>{children}</>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };

/**
 * Clean up attributes from the rendered component by removing the generated ids.
 *
 * Before:
 *   for="input-type-text-123456"
 *   id="input-type-text-123456"
 *
 * After:
 *   for="input-type-text"
 *   id="input-type-text"
 *
 * This is useful during snapshot testing. Snapshot comparisons will fail without addressing these items.
 *
 * @param {HTMLElement} container The component to clean up.
 * @param {prefixes}      string[]  The reference substrings to leave alone.
 */
