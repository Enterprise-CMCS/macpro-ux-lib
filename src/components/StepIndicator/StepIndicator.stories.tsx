import React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";
import { StepIndicator } from "./StepIndicator";

export default {
  title: "COMPONENTS/StepIndicator",
  component: StepIndicator,
  args: {
    steps: [
      { label: "Personal information", order: 1 },
      { label: "Household status", order: 2 },
      { label: "Supporting documents", order: 3 },
      { label: "Signature", order: 4 },
      { label: "Review and submit", order: 5 },
    ],
    currentProgress: 3,
  },
  argTypes: {
    currentProgress: {
      description: "Number tracker of which step the user is currently on.",
    },
    steps: {
      description:
        "An array of step object data which consist of a label and the order number. The order number is 1 based.",
    },
    counters: {
      description:
        "Boolean which shows the counters alternative to the step indidcator.",
    },
    smallCounters: {
      description:
        "Boolean which shows the smaller counters alternative to the step indicator. Only shows if counters is true.",
    },
    showLabels: {
      description: "Dictates whether or not labels are show under each step.",
    },
    isCentered: {
      description:
        "Dictates whether the labels are centered under the steps. Labels must be true.",
    },
  },
  parameters: {
    docs: {
      description: {
        component:
          "A step indicator updates users on their progress through a multi-step process.",
      },
    },
  },
} as ComponentMeta<typeof StepIndicator>;

const Template: ComponentStory<typeof StepIndicator> = ({ ...rest }) => (
  <StepIndicator {...rest} />
);

export const Default = Template.bind({});
Default.args = {};

export const NoLabels = Template.bind({});
NoLabels.args = { showLabels: false };

export const Centered = Template.bind({});
Centered.args = { isCentered: true };

export const Counters = Template.bind({});
Counters.args = { counters: true };

export const SmallCounters = Template.bind({});
SmallCounters.args = { counters: true, smallCounters: true };

export const CenteredCounters = Template.bind({});
CenteredCounters.args = { counters: true, isCentered: true };
