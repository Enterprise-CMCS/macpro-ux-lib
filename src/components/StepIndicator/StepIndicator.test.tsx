import React from "react";
import { screen, render } from "../../test-setup";
import { StepIndicator } from "./StepIndicator";

describe("Tests for the StepIndicator component", () => {
  const steps = [
    { label: "Personal information", order: 1 },
    { label: "Household status", order: 2 },
    { label: "Supporting documents", order: 3 },
    { label: "Signature", order: 4 },
    { label: "Review and submit", order: 5 },
  ];

  it("should render the StepIndicator component", () => {
    render(
      <StepIndicator
        data-testid="StepAlert"
        currentProgress={1}
        steps={steps}
      />
    );

    const StepComp = screen.getByTestId("StepAlert");

    expect(StepComp).toBeInTheDocument();
  });

  it("should render the StepIndicator component", () => {
    render(
      <StepIndicator
        data-testid="StepAlert"
        currentProgress={1}
        steps={steps}
        isCentered={true}
      />
    );

    const StepComp = screen.getByTestId("StepAlert");
    expect(StepComp).toBeInTheDocument();
  });

  it("should render the StepIndicator component", () => {
    render(
      <StepIndicator
        data-testid="StepAlert"
        currentProgress={3}
        steps={steps}
        counters={true}
        smallCounters={true}
        showLabels={false}
      />
    );

    const StepComp = screen.getByTestId("StepAlert");

    expect(StepComp).toHaveClass("usa-step-indicator--counters-sm");
  });

  it("should render the StepIndicator component", () => {
    render(
      <StepIndicator
        data-testid="StepAlert"
        currentProgress={5}
        steps={steps}
        counters={true}
        isCentered={true}
      />
    );
    const StepComp = screen.getByTestId("StepAlert");

    expect(StepComp).toHaveClass("usa-step-indicator--center");
    expect(StepComp).toHaveClass("usa-step-indicator--counters");
  });
});
