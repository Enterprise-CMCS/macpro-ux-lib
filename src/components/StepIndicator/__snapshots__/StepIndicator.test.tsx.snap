// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for the StepIndicator component compontent snapshots should render a info alert 1`] = `
<div>
  <div
    aria-label="progress"
    class="usa-step-indicator usa-step-indicator--center usa-step-indicator--counters"
    data-testid="StepAlert"
    role="progressbar"
  >
    <ol
      class="usa-step-indicator__segments"
    >
      <li
        class="usa-step-indicator__segment usa-step-indicator__segment--complete"
      >
        <span
          class="usa-step-indicator__segment-label"
        >
          Personal information
          <span
            class="usa-sr-only"
          >
            not completed
          </span>
        </span>
      </li>
      <li
        class="usa-step-indicator__segment usa-step-indicator__segment--complete"
      >
        <span
          class="usa-step-indicator__segment-label"
        >
          Household status
          <span
            class="usa-sr-only"
          >
            not completed
          </span>
        </span>
      </li>
      <li
        class="usa-step-indicator__segment usa-step-indicator__segment--complete"
      >
        <span
          class="usa-step-indicator__segment-label"
        >
          Supporting documents
          <span
            class="usa-sr-only"
          >
            not completed
          </span>
        </span>
      </li>
      <li
        class="usa-step-indicator__segment usa-step-indicator__segment--complete"
      >
        <span
          class="usa-step-indicator__segment-label"
        >
          Signature
          <span
            class="usa-sr-only"
          >
            not completed
          </span>
        </span>
      </li>
      <li
        class="usa-step-indicator__segment usa-step-indicator__segment--current"
      >
        <span
          class="usa-step-indicator__segment-label"
        >
          Review and submit
        </span>
      </li>
    </ol>
    <div
      class="usa-step-indicator__header"
    >
      <h4
        class="usa-step-indicator__heading"
      >
        <span
          class="usa-step-indicator__heading-counter"
        >
          <span
            class="usa-sr-only"
          >
            Step
          </span>
          <span
            class="usa-step-indicator__current-step margin-right-05"
          >
            5
          </span>
          <span
            class="usa-step-indicator__total-steps"
          >
            of 
            5
          </span>
           
        </span>
        <span
          class="usa-step-indicator__heading-text"
        >
          Test
        </span>
      </h4>
    </div>
  </div>
</div>
`;
